logger.c

```diff
@@ -0,0 +1,75 @@
+#include "logger.h"
+
+#define BUILD_BUG_ON_ZERO(expr) ((int)(sizeof(struct { int:(-!!(expr)); })))
+#define __same_type(a, b) __builtin_types_compatible_p(typeof(a), typeof(b))
+#define ARRAY_SIZE(arr) \
+        (BUILD_BUG_ON_ZERO(__same_type((arr), &(arr)[0])), (sizeof(arr) / sizeof((arr)[0])))
+#define LOGGER_OPTIONS_SIZE ARRAY_SIZE(logger)
+
+static const logger_pair_t logger[] = {
+  {
+    .log_name  = SERVER,
+    .log_value = LOG_KIND_SERVER,
+    .length    = sizeof(SERVER)
+  },
+  {
+    .log_name  = THERMOCOUPLE,
+    .log_value = LOG_KIND_THERMOCOUPLE,
+    .length    = sizeof(THERMOCOUPLE)
+  }
+};
+
+static const logger_pair_t*
+find_logger_by_name(const char* const name)
+{
+  const logger_pair_t* begin = &logger[0];
+  const logger_pair_t* const end = begin + LOGGER_OPTIONS_SIZE;
+
+  while (begin != end && strcmp(name, begin->log_name))
+   ++begin;
+
+  return begin;
+}
+
+void
+set_log(const char* buf, const unsigned set, uint8_t* log_bits)
+{
+  const logger_pair_t* const log = find_logger_by_name(buf);
+
+  if(log == &logger[0] + LOGGER_OPTIONS_SIZE)
+    return;
```

Comment:
Consider defining const for end "iterator". IMO It makes it slightly easier to read.

```C
const logger_pair_t * const log_end = &logger[0] + LOGGER_OPTIONS_SIZE;
if (log == log_end)
  return;
```